### PR TITLE
hardening/746_replace_getaccumulatorforrollback_with_clone

### PR DESCRIPTION
**Pull request by pcoello25**
* Fix issue #746 
* 100% unit test passed: 
```
Results :
Tests run: 88, Failures: 0, Errors: 0, Skipped: 0
```
* (unofficial) e2e tests passed:

I have tested a simple running of Cygnus with a `mysql-sink`. Once started, mysql was stopped and a notification was sent in order to check rollbacking. After first rollbacking, mysql was started again and persistence was successfull.

Receiving data:
```
time=2016-02-10T11:25:26.136CET | lvl=INFO | trans=1455099916-77-0000000000 | srv=fiwareservice | subsrv=fiwareservicepath | function=getEvents | comp=Cygnus | msg=com.telefonica.iot.cygnus.handlers.OrionRestHandler[250] : Starting transaction (1455099916-77-0000000000)
time=2016-02-10T11:25:26.137CET | lvl=INFO | trans=1455099916-77-0000000000 | srv=fiwareservice | subsrv=fiwareservicepath | function=getEvents | comp=Cygnus | msg=com.telefonica.iot.cygnus.handlers.OrionRestHandler[277] : Received data ({    "subscriptionId" : "51c0ac9ed714fb3b37d7d5a8",    "originator" : "localhost",    "contextResponses" : [        {            "contextElement" : {                "attributes" : [                    {                        "name" : "capacity",                        "type" : "float",                       "value" : "12.20"                    }                ],                "type" : "Room",                "isPattern" : "false",                "id" : "Room1"            },            "statusCode" : {                "code" : "200",                "reasonPhrase" : "OK"            }        }    ]})
```

Persistence error (due to mysql is stopped) and rollbacking:
```
time=2016-02-10T11:25:42.766CET | lvl=INFO | trans=1455099916-77-0000000000 | srv=fiwareservice | subsrv=fiwareservicepath | function=persistAggregation | comp=Cygnus | msg=com.telefonica.iot.cygnus.sinks.OrionMySQLSink[457] : [mysql-sink] Persisting data at OrionMySQLSink. Database (fiwareservice), Table (fiwareservicepath_room1_room), Fields ((recvTimeTs,recvTime,fiwareservicepath,entityId,entityType,attrName,attrType,attrValue,attrMd)), Values (('1455099926140','2016-02-10T10:25:26.140Z','fiwareservicepath','Room1','Room','capacity','float','12.20','[]'))
time=2016-02-10T11:25:42.829CET | lvl=DEBUG | trans=1455099916-77-0000000000 | srv=fiwareservice | subsrv=fiwareservicepath | function=createConnection | comp=Cygnus | msg=com.telefonica.iot.cygnus.backends.mysql.MySQLBackendImpl$MySQLDriver[267] : Connecting to jdbc:mysql://localhost:3306/?user=root&password=XXXXXXXXXX
time=2016-02-10T11:25:42.958CET | lvl=DEBUG | trans=1455099916-77-0000000000 | srv=fiwareservice | subsrv=fiwareservicepath | function=processNewBatches | comp=Cygnus | msg=com.telefonica.iot.cygnus.sinks.OrionSink[449] : [com.telefonica.iot.cygnus.backends.mysql.MySQLBackendImpl$MySQLDriver.getConnection(MySQLBackendImpl.java:230), com.telefonica.iot.cygnus.backends.mysql.MySQLBackendImpl.createDatabase(MySQLBackendImpl.java:78), com.telefonica.iot.cygnus.sinks.OrionMySQLSink.persistAggregation(OrionMySQLSink.java:464), com.telefonica.iot.cygnus.sinks.OrionMySQLSink.persistBatch(OrionMySQLSink.java:179), com.telefonica.iot.cygnus.sinks.OrionSink.processNewBatches(OrionSink.java:439), com.telefonica.iot.cygnus.sinks.OrionSink.process(OrionSink.java:183), org.apache.flume.sink.DefaultSinkProcessor.process(DefaultSinkProcessor.java:68), org.apache.flume.SinkRunner$PollingRunner.run(SinkRunner.java:147), java.lang.Thread.run(Thread.java:745)]
time=2016-02-10T11:25:42.959CET | lvl=ERROR | trans=1455099916-77-0000000000 | srv=fiwareservice | subsrv=fiwareservicepath | function=processNewBatches | comp=Cygnus | msg=com.telefonica.iot.cygnus.sinks.OrionSink[453] : Persistence error (Communications link failure

The last packet sent successfully to the server was 0 milliseconds ago. The driver has not received any packets from the server.)
time=2016-02-10T11:25:42.959CET | lvl=INFO | trans=1455099916-77-0000000000 | srv=fiwareservice | subsrv=fiwareservicepath | function=processNewBatches | comp=Cygnus | msg=com.telefonica.iot.cygnus.sinks.OrionSink[454] : Rollbacking (1455099916-77-0000000000)
```

Rollbacking again:
```
time=2016-02-10T11:25:46.961CET | lvl=INFO | trans=1455099916-77-0000000000 | srv=fiwareservice | subsrv=fiwareservicepath | function=persistAggregation | comp=Cygnus | msg=com.telefonica.iot.cygnus.sinks.OrionMySQLSink[457] : [mysql-sink] Persisting data at OrionMySQLSink. Database (fiwareservice), Table (fiwareservicepath_room1_room), Fields ((recvTimeTs,recvTime,fiwareservicepath,entityId,entityType,attrName,attrType,attrValue,attrMd)), Values (('1455099926140','2016-02-10T10:25:26.140Z','fiwareservicepath','Room1','Room','capacity','float','12.20','[]'))
time=2016-02-10T11:25:46.962CET | lvl=DEBUG | trans=1455099916-77-0000000000 | srv=fiwareservice | subsrv=fiwareservicepath | function=createConnection | comp=Cygnus | msg=com.telefonica.iot.cygnus.backends.mysql.MySQLBackendImpl$MySQLDriver[267] : Connecting to jdbc:mysql://localhost:3306/?user=root&password=XXXXXXXXXX
time=2016-02-10T11:25:46.966CET | lvl=DEBUG | trans=1455099916-77-0000000000 | srv=fiwareservice | subsrv=fiwareservicepath | function=processRollbackedBatches | comp=Cygnus | msg=com.telefonica.iot.cygnus.sinks.OrionSink[336] : [com.telefonica.iot.cygnus.backends.mysql.MySQLBackendImpl$MySQLDriver.getConnection(MySQLBackendImpl.java:230), com.telefonica.iot.cygnus.backends.mysql.MySQLBackendImpl.createDatabase(MySQLBackendImpl.java:78), com.telefonica.iot.cygnus.sinks.OrionMySQLSink.persistAggregation(OrionMySQLSink.java:464), com.telefonica.iot.cygnus.sinks.OrionMySQLSink.persistBatch(OrionMySQLSink.java:179), com.telefonica.iot.cygnus.sinks.OrionSink.processRollbackedBatches(OrionSink.java:330), com.telefonica.iot.cygnus.sinks.OrionSink.process(OrionSink.java:185), org.apache.flume.sink.DefaultSinkProcessor.process(DefaultSinkProcessor.java:68), org.apache.flume.SinkRunner$PollingRunner.run(SinkRunner.java:147), java.lang.Thread.run(Thread.java:745)]
time=2016-02-10T11:25:46.967CET | lvl=ERROR | trans=1455099916-77-0000000000 | srv=fiwareservice | subsrv=fiwareservicepath | function=processRollbackedBatches | comp=Cygnus | msg=com.telefonica.iot.cygnus.sinks.OrionSink[340] : Persistence error (Communications link failure

The last packet sent successfully to the server was 0 milliseconds ago. The driver has not received any packets from the server.)
time=2016-02-10T11:25:46.967CET | lvl=INFO | trans=1455099916-77-0000000000 | srv=fiwareservice | subsrv=fiwareservicepath | function=processRollbackedBatches | comp=Cygnus | msg=com.telefonica.iot.cygnus.sinks.OrionSink[341] : Rollbacking again (1455099916-77-0000000000)
```

Starting mysql again:
```
time=2016-02-10T11:25:51.970CET | lvl=INFO | trans=1455099916-77-0000000000 | srv=fiwareservice | subsrv=fiwareservicepath | function=persistAggregation | comp=Cygnus | msg=com.telefonica.iot.cygnus.sinks.OrionMySQLSink[457] : [mysql-sink] Persisting data at OrionMySQLSink. Database (fiwareservice), Table (fiwareservicepath_room1_room), Fields ((recvTimeTs,recvTime,fiwareservicepath,entityId,entityType,attrName,attrType,attrValue,attrMd)), Values (('1455099926140','2016-02-10T10:25:26.140Z','fiwareservicepath','Room1','Room','capacity','float','12.20','[]'))
time=2016-02-10T11:25:51.971CET | lvl=DEBUG | trans=1455099916-77-0000000000 | srv=fiwareservice | subsrv=fiwareservicepath | function=createConnection | comp=Cygnus | msg=com.telefonica.iot.cygnus.backends.mysql.MySQLBackendImpl$MySQLDriver[267] : Connecting to jdbc:mysql://localhost:3306/?user=root&password=XXXXXXXXXX
time=2016-02-10T11:25:52.204CET | lvl=DEBUG | trans=1455099916-77-0000000000 | srv=fiwareservice | subsrv=fiwareservicepath | function=createDatabase | comp=Cygnus | msg=com.telefonica.iot.cygnus.backends.mysql.MySQLBackendImpl[88] : Executing MySQL query 'create database if not exists `fiwareservice`'
time=2016-02-10T11:25:52.208CET | lvl=DEBUG | trans=1455099916-77-0000000000 | srv=fiwareservice | subsrv=fiwareservicepath | function=createConnection | comp=Cygnus | msg=com.telefonica.iot.cygnus.backends.mysql.MySQLBackendImpl$MySQLDriver[267] : Connecting to jdbc:mysql://localhost:3306/fiwareservice?user=root&password=XXXXXXXXXX
time=2016-02-10T11:25:52.214CET | lvl=DEBUG | trans=1455099916-77-0000000000 | srv=fiwareservice | subsrv=fiwareservicepath | function=createTable | comp=Cygnus | msg=com.telefonica.iot.cygnus.backends.mysql.MySQLBackendImpl[118] : Executing MySQL query 'create table if not exists `fiwareservicepath_room1_room` (recvTimeTs long,recvTime text,fiwareservicepath text,entityId text,entityType text,attrName text,attrType text,attrValue text,attrMd text)'
time=2016-02-10T11:25:52.217CET | lvl=DEBUG | trans=1455099916-77-0000000000 | srv=fiwareservice | subsrv=fiwareservicepath | function=createConnection | comp=Cygnus | msg=com.telefonica.iot.cygnus.backends.mysql.MySQLBackendImpl$MySQLDriver[267] : Connecting to jdbc:mysql://localhost:3306/fiwareservice?user=root&password=XXXXXXXXXX
time=2016-02-10T11:25:52.224CET | lvl=DEBUG | trans=1455099916-77-0000000000 | srv=fiwareservice | subsrv=fiwareservicepath | function=insertContextData | comp=Cygnus | msg=com.telefonica.iot.cygnus.backends.mysql.MySQLBackendImpl[143] : Executing MySQL query 'insert into `fiwareservicepath_room1_room` (recvTimeTs,recvTime,fiwareservicepath,entityId,entityType,attrName,attrType,attrValue,attrMd) values ('1455099926140','2016-02-10T10:25:26.140Z','fiwareservicepath','Room1','Room','capacity','float','12.20','[]')'
time=2016-02-10T11:25:52.262CET | lvl=INFO | trans=1455099916-77-0000000000 | srv=fiwareservice | subsrv=fiwareservicepath | function=processRollbackedBatches | comp=Cygnus | msg=com.telefonica.iot.cygnus.sinks.OrionSink[331] : Finishing transaction (1455099916-77-0000000000)
```

mysql databases: 
```
mysql> show databases;
+--------------------+
| Database           |
+--------------------+
| information_schema |
| fiwareservice      |
| mysql              |
| performance_schema |
| sys                |
+--------------------+
5 rows in set (0,02 sec)

mysql> use fiwareservice
Reading table information for completion of table and column names
You can turn off this feature to get a quicker startup with -A

Database changed

mysql> show tables;
+------------------------------+
| Tables_in_fiwareservice      |
+------------------------------+
| fiwareservicepath_room1_room |
+------------------------------+
1 row in set (0,00 sec)

mysql> SELECT * FROM fiwareservicepath_room1_room;
+---------------+--------------------------+-------------------+----------+------------+----------+----------+-----------+--------+
| recvTimeTs    | recvTime                 | fiwareservicepath | entityId | entityType | attrName | attrType | attrValue | attrMd |
+---------------+--------------------------+-------------------+----------+------------+----------+----------+-----------+--------+
| 1455099926140 | 2016-02-10T10:25:26.140Z | fiwareservicepath | Room1    | Room       | capacity | float    | 12.20     | []     |
+---------------+--------------------------+-------------------+----------+------------+----------+----------+-----------+--------+
1 row in set (0,00 sec)
```

* Assignee: @frbattid

### CODE REVIEW

File: src/main/java/com/telefonica/iot/cygnus/sinks/OrionSink.java [hunk 1]
```diff
@@ -733,29 +733,16 @@ public void initialize(long startDateMs) {
...
-
-            return accumulatorForRollback;
-        } // getAccumulatorForRollback
+        @Override
+        public Accumulator clone() {
+            try {
+                Accumulator acc = (Accumulator) super.clone();
+                return acc; 
+            }
+            catch (CloneNotSupportedException ce) {
```
Comment (frbattid) [inline src/main/java/com/telefonica/iot/cygnus/sinks/OrionSink.java]: `catch` must be after the `{`. i.e.:

    try {
       ...
    } catch {
       ...
    }

File: CHANGES_NEXT_RELEASE [hunk 2]
```diff
@@ -4,3 +4,4 @@
 - [BUG] Fix wrong spec date regarding release/0.12.0 (#750)
 - [HARDENING] Add check for null content-type in notifications (#743) 
 - [HARDENING] Add support for configurable number of partitions and replication factor in OrionKafkaSink (#736)
+- [HARDENING] Replace method getAccumulatorForRollBack with clone (#746)
```
Comment (frbattid) [inline CHANGES_NEXT_RELEASE]: Please, specify the class, i.e.:

"Replace method Accumulator.getAccumulatorForRollback() with Accumulator.clone() (#746)"

File: src/main/java/com/telefonica/iot/cygnus/sinks/OrionSink.java (hunk 1 above)
Comment (pcoello25) [inline src/main/java/com/telefonica/iot/cygnus/sinks/OrionSink.java]: Fixed in b141680

File: CHANGES_NEXT_RELEASE (hunk 2 above)
Comment (pcoello25) [inline CHANGES_NEXT_RELEASE]: Fixed in b141680

**Comment by frbattid** [comment]
LGTM

**Pull request merged**